cli/module_build.go

```diff
@@ -437,13 +437,81 @@ func reloadModuleAction(c *cli.Context, vc *viamClient) error {
 				return err
 			}
 		}
+		if !c.Bool(moduleFlagLocal) {
+			if manifest == nil || manifest.Build == nil || manifest.Build.Path == "" {
+				return errors.New(
+					"remote reloading requires a meta.json with the 'build.path' field set. " +
+						"try --local if you are testing on the same machine.",
+				)
+			}
+			if err := validateReloadableArchive(c, manifest.Build); err != nil {
+				return err
+			}
+			if err := addShellService(c, vc, part.Part, true); err != nil {
+				return err
+			}
+			infof(c.App.Writer, "copying %s to part %s", manifest.Build.Path, part.Part.Id)
+			err = vc.copyFilesToFqdn(
+				part.Part.Fqdn, c.Bool(debugFlag), false, false, []string{manifest.Build.Path},
+				reloadingDestination(c, manifest), logging.NewLogger("reload"))
+			if err != nil {
+				return err
+			}
+		}
 		needsRestart, err = configureModule(c, vc, manifest, part.Part)
 		if err != nil {
 			return err
 		}
 	}
 	if needsRestart {
-		return restartModule(c, vc, part.Part, manifest)
+		err = restartModule(c, vc, part.Part, manifest)
+		if s, ok := status.FromError(err); ok && s.Code() == codes.NotFound {
+			warningf(c.App.ErrWriter,
+				"viam-server couldn't find your module to restart it; "+
+					"this happens if the module couldn't start, and may not indicate an error.",
+			)
+		}
+		return err
+	}
+	infof(c.App.Writer, "Reload complete")
+	return nil
+}
+
+// this chooses a destination path for the module archive.
+func reloadingDestination(c *cli.Context, manifest *moduleManifest) string {
+	return filepath.Join(c.String(moduleFlagHomeDir),
+		".viam", config.PackagesDirName+config.LocalPackagesSuffix,
+		utils.SanitizePath(localizeModuleID(manifest.ModuleID)+"-"+manifest.Build.Path))
+}
+
+// validateReloadableArchive returns an error if there is a fatal issue (for now just file not found).
+// It also logs warnings for likely problems.
+func validateReloadableArchive(c *cli.Context, build *manifestBuildInfo) error {
+	reader, err := os.Open(build.Path)
+	if err != nil {
+		return err
+	}
+	decompressed, err := gzip.NewReader(reader)
+	if err != nil {
+		return err
+	}
+	archive := tar.NewReader(decompressed)
+	metaFound := false
+	for {
+		header, err := archive.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return errors.Wrapf(err, "reading tar at %s", build.Path)
+		}
+		if header.Name == "meta.json" {
+			metaFound = true
+			break
+		}
+	}
+	if !metaFound {
+		warningf(c.App.ErrWriter, "archive at %s doesn't contain a meta.json, your module will probably fail to start", build.Path)
```

Comment:
This isn't an actual error? Just a warning?

Reply:
I didn't want to couple it too closely to the meta-in-tarball behavior, plus this is a developer feature with quick feedback if something is wrong

no objection to returning an error if you think it's cleaner

Reply:
> so I prioritized letting the user break things over rejecting likely breakages.

Seems like fair reasoning to me; thanks for the context.